### PR TITLE
Display exception notes in tracebacks

### PR DESCRIPTION
**Pull request by EtiennePelletier**
[PEP 678](https://peps.python.org/pep-0678/) introduced the ability to add notes to exception objects. This has been [released in Python 3.11](https://docs.python.org/3/library/exceptions.html#BaseException.add_note) and is currently not implemented in IPython. These changes are fully compatible with older Python versions that don't include PEP 678.

Here's a sample test that shows the consistency in Python's stdlib traceback module (test 1) and the difference between Python and IPython's runtimes (test 2):
```python
import traceback

print('--- test 1 ---')
try:
    raise Exception('Testing notes')
except Exception as e:
    e.add_note('Does this work?')
    e.add_note('Yes!')
    traceback.print_exc()

print('\n--- test 2 ---')
try:
    raise Exception('Testing notes')
except Exception as e:
    e.add_note('Does this work?')
    e.add_note('No!')
    raise
```

When executed with Python 3.11, both notes are displayed in both tracebacks:
```
$ python test.py 
--- test 1 ---
Traceback (most recent call last):
  File "/app/test.py", line 5, in <module>
    raise Exception('Testing notes')
Exception: Testing notes
Does this work?
Yes!

--- test 2 ---
Traceback (most recent call last):
  File "/app/test.py", line 13, in <module>
    raise Exception('Testing notes')
Exception: Testing notes
Does this work?
No!
```

In IPython's VerboseTB does not yet handle exception notes:
```
$ ipython test.py 
--- test 1 ---
Traceback (most recent call last):
  File "/app/test.py", line 5, in <module>
    raise Exception('Testing notes')
Exception: Testing notes
Does this work?
Yes!

--- test 2 ---
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
File /app/test.py:13
     11 print('\n--- test 2 ---')
     12 try:
---> 13     raise Exception('Testing notes')
     14 except Exception as e:
     15     e.add_note('Does this work?')

Exception: Testing notes
```

The changes I am suggesting are inspired from implementation of [Lib/traceback.py](https://github.com/python/cpython/blob/main/Lib/traceback.py) (search for `__notes__`) and improvements for dealing with edge cases more nicely in [cpython#103897](https://github.com/python/cpython/pull/103897).

Although notes are meant to be strings only, I kept some inspiration from the existing exception handling to ensure that the notes are uncolored and bytes decoded, if there are any. I am definitely open to using a different color if deemed better. For context, `bpython` keeps the notes uncolored, and [Python's tutorial](https://docs.python.org/3/tutorial/errors.html#enriching-exceptions-with-notes) puts them in light gray, like the line numbers.

Here's how the test 2 looks like after these changes:
image

## :snake: :man_juggling:

### CODE REVIEW

**Comment by Zac-HD** [comment]
Oops, I checked for existing issues a few hours before this was opened and then created https://github.com/ipython/ipython/pull/14040 - it has tests and a changelog, but is a bit less careful about edge cases.  Steal whatever you want and then I'll close it?

**Comment by Carreau** [comment]
I'm guessing you both are at PyCon sprints !? Enjoy ! I'll likely get that in during May :-).

**Comment by EtiennePelletier** [comment]
Changes cherry-picked, good catch Zac on the Plain traceback format!

File: .gitignore [hunk 1]
```diff
@@ -27,6 +27,7 @@ __pycache__
 *.swp
 .pytest_cache
 .python-version
+.venv*/
```
Comment (Carreau) [inline .gitignore:30]: You likely want that in your [global gitignore](https://gist.github.com/subfuzion/db7f57fff2fb6998a16c) to not have to do it for all your repos.

**Comment by Carreau** [comment]
Let's try this. Sorry I was mostly travelling last month.
